src/test/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecutionTest.java

```diff
@@ -121,4 +134,64 @@ public void stop(@Nonnull Throwable cause) throws Exception {
 
     }
 
+    @Issue("JENKINS-34637")
+    @Test public void currentExecutions() throws Exception {
+        WorkflowJob p = jenkins.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition("parallel main: {retainsBody {parallel a: {retainsBody {semaphore 'a'}}, b: {retainsBody {semaphore 'b'}}}}, aside: {semaphore 'c'}", true));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("a/1", b);
+        SemaphoreStep.waitForStart("b/1", b);
+        SemaphoreStep.waitForStart("c/1", b);
+        final RetainsBodyStep.Execution[] execs = new RetainsBodyStep.Execution[3];
+        StepExecution.applyAll(RetainsBodyStep.Execution.class, new Function<RetainsBodyStep.Execution, Void>() {
+            @Override public Void apply(RetainsBodyStep.Execution exec) {
+                execs[exec.count] = exec;
+                return null;
+            }
+        }).get();
+        assertNotNull(execs[0]);
+        assertNotNull(execs[1]);
+        assertNotNull(execs[2]);
+        final Set<SemaphoreStep.Execution> semaphores = new HashSet<>();
+        StepExecution.applyAll(SemaphoreStep.Execution.class, new Function<SemaphoreStep.Execution, Void>() {
+            @Override public Void apply(SemaphoreStep.Execution exec) {
+                if (exec.getStatus().matches("waiting on [ab]/1")) {
+                    semaphores.add(exec);
+                }
+                return null;
+            }
+        }).get();
+        assertThat(semaphores, Matchers.<SemaphoreStep.Execution>iterableWithSize(2));
+        Collection<StepExecution> currentExecutions1 = execs[1].body.getCurrentExecutions(); // A or B, does not matter
+        assertThat(/* irritatingly, iterableWithSize does not show the collection in its mismatch message */currentExecutions1.toString(),
+            currentExecutions1, Matchers.<StepExecution>iterableWithSize(1));
+        Collection<StepExecution> currentExecutions2 = execs[2].body.getCurrentExecutions();
+        assertThat(currentExecutions2, Matchers.<StepExecution>iterableWithSize(1));
+        assertEquals(semaphores, Sets.union(Sets.newLinkedHashSet(currentExecutions1), Sets.newLinkedHashSet(currentExecutions2)));
+        assertEquals(semaphores, Sets.newLinkedHashSet(execs[0].body.getCurrentExecutions())); // the top-level one
+        execs[0].body.cancel();
+        SemaphoreStep.success("c/1", null);
+        jenkins.assertBuildStatus(Result.ABORTED, jenkins.waitForCompletion(b));
```

Comment:
Failed before fix: called `ParallelStepExecution.stop`, which delegated to `RetainsBodyStep.Execution.stop`, rather than going straight to `SemaphoreStep.Execution.stop`.